From c84c0ea9f7bb457f6f3f28bf0cd3280e7dae1915 Mon Sep 17 00:00:00 2001
From: Antonio Borneo <antonio.borneo@foss.st.com>
Date: Tue, 4 Feb 2025 11:59:52 +0100
Subject: [PATCH 1590/1695] plat-stm32mp2: let PWR IRQ to propagate to parent
 GIC

Populate all the API in struct itr_ops to propagate the requests
to the parent interrupt controller (GIC).
The struct itr_ops will be used later.

While there:
- rename gic_hdl with the more generic parent_hdl;
- get parent interrupt controller from DT;
- replace interrupt_alloc_add_handler() with interrupt_create_handler();
- use mask/unmask in interrupt context instead of enable/disable.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git ce030878830486cd771d310bd7585694a91faeda plat-stm32mp2: let PWR IRQ to propagate to parent GIC]
Signed-off-by: Antonio Borneo <antonio.borneo@foss.st.com>
Change-Id: I4b1534ab3c42dba4331f8a172ccacbef824739df
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/434935
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
---
 .../arm/plat-stm32mp2/drivers/stm32mp25_pwr.c |  13 +-
 .../arm/plat-stm32mp2/drivers/stm32mp25_pwr.h |   3 +-
 .../plat-stm32mp2/drivers/stm32mp25_pwr_irq.c | 238 +++++++++++++++++-
 3 files changed, 231 insertions(+), 23 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr.c b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr.c
index fd2f38808..ee0a5113c 100644
--- a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr.c
+++ b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr.c
@@ -85,7 +85,6 @@
 
 struct pwr_pdata {
 	vaddr_t base;
-	int interrupt;
 	uint8_t nb_ressources;
 	struct rif_conf_data *conf_data;
 };
@@ -278,7 +277,7 @@ static void parse_dt(const void *fdt, int node)
 	cuint = fdt_getprop(fdt, node, "st,protreg", &lenp);
 	if (!cuint) {
 		DMSG("No RIF configuration available");
-		goto skip_rif;
+		return;
 	}
 
 	pwr_d->nb_ressources = (unsigned int)(lenp / sizeof(uint32_t));
@@ -300,14 +299,6 @@ static void parse_dt(const void *fdt, int node)
 	for (i = 0; i < pwr_d->nb_ressources; i++)
 		stm32_rif_parse_cfg(fdt32_to_cpu(cuint[i]), pwr_d->conf_data,
 				    _PWR_NB_RESOURCES);
-
-skip_rif:
-#ifdef CFG_STM32_PWR_IRQ
-	if (info.interrupt == DT_INFO_INVALID_INTERRUPT)
-		panic("No interrupt defined in PWR");
-
-	pwr_d->interrupt = info.interrupt;
-#endif
 }
 
 static void pm_resume(void)
@@ -404,7 +395,7 @@ static TEE_Result stm32mp_pwr_probe(const void *fdt, int node,
 		panic("Failed to apply rif_config");
 
 #ifdef CFG_STM32_PWR_IRQ
-	res = stm32mp25_pwr_irq_probe(fdt, node, pwr_d->interrupt);
+	res = stm32mp25_pwr_irq_probe(fdt, node);
 	if (res) {
 		if (pwr_d->conf_data) {
 			free(pwr_d->conf_data->access_mask);
diff --git a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr.h b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr.h
index 4985a8db7..ebe19996d 100644
--- a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr.h
+++ b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr.h
@@ -47,8 +47,7 @@ stm32mp25_pwr_itr_alloc_add(const void *fdt, int wp_node, size_t it,
 void stm32mp25_pwr_itr_enable(size_t it);
 void stm32mp25_pwr_itr_disable(size_t it);
 
-TEE_Result
-stm32mp25_pwr_irq_probe(const void *fdt, int node, int interrupt);
+TEE_Result stm32mp25_pwr_irq_probe(const void *fdt, int node);
 #endif /* CFG_STM32_PWR_IRQ */
 
 vaddr_t stm32_pwr_base(void);
diff --git a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_irq.c b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_irq.c
index 904440b8a..2bb5591c6 100644
--- a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_irq.c
+++ b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_irq.c
@@ -51,7 +51,7 @@ enum wkup_pull_setting {
 struct pwr_wkup_data {
 	struct stm32_pinctrl_list *pinctrl_list;
 	struct itr_handler *hdl[PWR_NB_WAKEUPPINS];
-	struct itr_handler *gic_hdl;
+	struct itr_handler *parent_hdl;
 	struct stm32_exti_pdata *exti;
 	bool threaded[PWR_NB_WAKEUPPINS];
 	bool pending[PWR_NB_WAKEUPPINS];
@@ -150,7 +150,7 @@ static enum itr_return pwr_it_handler(struct itr_handler *handler)
 
 	VERBOSE_PWR("it_handler");
 
-	interrupt_disable(interrupt_get_main_chip(), priv->gic_hdl->it);
+	interrupt_mask(priv->parent_hdl->chip, priv->parent_hdl->it);
 
 	for (i = 0; i < PWR_NB_WAKEUPPINS; i++) {
 		uint32_t wkupcr = 0;
@@ -171,7 +171,7 @@ static enum itr_return pwr_it_handler(struct itr_handler *handler)
 		}
 	}
 
-	interrupt_enable(interrupt_get_main_chip(), priv->gic_hdl->it);
+	interrupt_unmask(priv->parent_hdl->chip, priv->parent_hdl->it);
 
 	return ITRR_HANDLED;
 }
@@ -280,6 +280,220 @@ void stm32mp25_pwr_itr_disable(size_t it)
 	cpu_spin_unlock_xrestore(&priv->spinlock, exceptions);
 }
 
+/* Register and configure an interrupt */
+static void stm32mp25_pwr_op_add(struct itr_chip *chip __unused,
+				 size_t it __unused, uint32_t type __unused,
+				 uint32_t prio __unused)
+{
+	/* TODO: this function is mandatory but not used. Will be removed! */
+	panic();
+}
+
+/* Enable an interrupt */
+static void stm32mp25_pwr_op_enable(struct itr_chip *chip __unused, size_t it)
+{
+	struct pwr_wkup_data *priv = pwr_wkup_d;
+	uint32_t exceptions = 0;
+
+	exceptions = cpu_spin_lock_xsave(&priv->spinlock);
+	stm32mp25_pwr_itr_enable_nolock(it);
+	cpu_spin_unlock_xrestore(&priv->spinlock, exceptions);
+
+	interrupt_enable(priv->parent_hdl->chip, priv->parent_hdl->it);
+}
+
+/* Disable an interrupt */
+static void stm32mp25_pwr_op_disable(struct itr_chip *chip __unused, size_t it)
+{
+	struct pwr_wkup_data *priv = pwr_wkup_d;
+	uint32_t exceptions = 0;
+
+	exceptions = cpu_spin_lock_xsave(&priv->spinlock);
+	stm32mp25_pwr_itr_disable_nolock(it);
+	cpu_spin_unlock_xrestore(&priv->spinlock, exceptions);
+
+	interrupt_disable(priv->parent_hdl->chip, priv->parent_hdl->it);
+}
+
+/* Mask an interrupt, may be called from an interrupt context */
+static void stm32mp25_pwr_op_mask(struct itr_chip *chip __unused, size_t it)
+{
+	struct pwr_wkup_data *priv = pwr_wkup_d;
+	uint32_t exceptions = 0;
+
+	exceptions = cpu_spin_lock_xsave(&priv->spinlock);
+	stm32mp25_pwr_itr_disable_nolock(it);
+	cpu_spin_unlock_xrestore(&priv->spinlock, exceptions);
+
+	interrupt_mask(priv->parent_hdl->chip, priv->parent_hdl->it);
+}
+
+/* Unmask an interrupt, may be called from an interrupt context */
+static void stm32mp25_pwr_op_unmask(struct itr_chip *chip __unused, size_t it)
+{
+	struct pwr_wkup_data *priv = pwr_wkup_d;
+	uint32_t exceptions = 0;
+
+	exceptions = cpu_spin_lock_xsave(&priv->spinlock);
+	stm32mp25_pwr_itr_enable_nolock(it);
+	cpu_spin_unlock_xrestore(&priv->spinlock, exceptions);
+
+	interrupt_unmask(priv->parent_hdl->chip, priv->parent_hdl->it);
+}
+
+/* Raise per-cpu interrupt (optional) */
+static void stm32mp25_pwr_op_raise_pi(struct itr_chip *chip __unused,
+				      size_t it __unused)
+{
+	struct pwr_wkup_data *priv = pwr_wkup_d;
+
+	/* nothing to do here, only forward to parent */
+
+	if (interrupt_can_raise_pi(priv->parent_hdl->chip))
+		interrupt_raise_pi(priv->parent_hdl->chip,
+				   priv->parent_hdl->it);
+}
+
+/* Raise a SGI (optional) */
+static void stm32mp25_pwr_op_raise_sgi(struct itr_chip *chip __unused,
+				       size_t it __unused, uint32_t cpu_mask)
+{
+	struct pwr_wkup_data *priv = pwr_wkup_d;
+
+	/* nothing to do here, only forward to parent */
+
+	if (interrupt_can_raise_sgi(priv->parent_hdl->chip))
+		interrupt_raise_sgi(priv->parent_hdl->chip,
+				    priv->parent_hdl->it, cpu_mask);
+}
+
+/* Set interrupt/cpu affinity (optional) */
+static void stm32mp25_pwr_op_set_affinity(struct itr_chip *chip __unused,
+					  size_t it __unused, uint8_t cpu_mask)
+{
+	struct pwr_wkup_data *priv = pwr_wkup_d;
+
+	/* nothing to do here, only forward to parent */
+
+	if (interrupt_can_set_affinity(priv->parent_hdl->chip))
+		interrupt_set_affinity(priv->parent_hdl->chip,
+				       priv->parent_hdl->it, cpu_mask);
+}
+
+/* Enable/disable power-management wake-on of an interrupt (optional) */
+static void stm32mp25_pwr_op_set_wake(struct itr_chip *chip __unused,
+				      size_t it __unused, bool on)
+{
+	struct pwr_wkup_data *priv = pwr_wkup_d;
+
+	/*
+	 * TODO:
+	 * Today, even with clients that call interrupt_set_wake(),
+	 * this driver incorrectly set the wakeup in WKUPENCPU1 as part of
+	 * enable() and unmask().
+	 * This driver must be reworked to separate:
+	 * - interrupt handling (simply forward to GIC);
+	 * - wakeup-source from client to handle WKUPENCPU1.
+	 * Client drivers should be reworked to request the wakeup.
+	 * For the moment, only forward to parent.
+	 */
+
+	if (interrupt_can_set_wake(priv->parent_hdl->chip))
+		interrupt_set_wake(priv->parent_hdl->chip,
+				   priv->parent_hdl->it, on);
+}
+
+static const struct itr_ops stm32mp2_pwr_itr_ops = {
+	.add		= stm32mp25_pwr_op_add,
+	.enable		= stm32mp25_pwr_op_enable,
+	.disable	= stm32mp25_pwr_op_disable,
+	.mask		= stm32mp25_pwr_op_mask,
+	.unmask		= stm32mp25_pwr_op_unmask,
+	.raise_pi	= stm32mp25_pwr_op_raise_pi,
+	.raise_sgi	= stm32mp25_pwr_op_raise_sgi,
+	.set_affinity	= stm32mp25_pwr_op_set_affinity,
+	.set_wake	= stm32mp25_pwr_op_set_wake,
+};
+DECLARE_KEEP_PAGER(stm32mp2_pwr_itr_ops);
+
+static struct itr_chip stm32mp2_pwr_itr_chip = {
+	.ops = &stm32mp2_pwr_itr_ops,
+	.name = "stm32mp2-pwr-irq",
+};
+
+__unused
+static TEE_Result stm32mp2_pwr_itr_dt_get(struct dt_pargs *args,
+					  void *priv_data,
+					  struct itr_desc *itr_desc_p)
+{
+	struct pwr_wkup_data *priv = priv_data;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned int trigger_mode = 0;
+	unsigned int itr_num = 0;
+	struct gpio *gpio = NULL;
+	unsigned int bank = 0;
+	unsigned int pin = 0;
+
+	VERBOSE_PWR("Pwr IRQ add");
+
+	assert(priv && args->args_count == 2);
+
+	itr_num = args->args[0];
+	trigger_mode = args->args[1];
+
+	assert(itr_num < PWR_NB_WAKEUPPINS);
+
+	switch (trigger_mode) {
+	case IRQ_TYPE_EDGE_RISING:
+		res = stm32_pwr_irq_set_trig(itr_num, PWR_WKUP_FLAG_RISING);
+		break;
+	case IRQ_TYPE_EDGE_FALLING:
+		res = stm32_pwr_irq_set_trig(itr_num, PWR_WKUP_FLAG_FALLING);
+		break;
+	default:
+		EMSG("Unsupported flags %#x", trigger_mode);
+		res = TEE_ERROR_BAD_PARAMETERS;
+		break;
+	}
+	if (res)
+		return res;
+
+	res = gpio_dt_get_by_index(args->fdt, args->phandle_node, itr_num,
+				   "wakeup", &gpio);
+	if (res)
+		return res;
+
+	bank = stm32_gpio_chip_bank_id(gpio->chip);
+	pin = gpio->pin;
+	if (bank != pin_map[itr_num].bank || pin != pin_map[itr_num].pin) {
+		EMSG("Invalid PWR WKUP%d on GPIO%c%"PRIu8" expected GPIO%c%"
+		     PRIu8, itr_num + 1,
+		     GPIO_PORT(bank), pin, GPIO_PORT(pin_map[itr_num].bank),
+		     pin_map[itr_num].pin);
+		panic();
+	}
+
+	/* Use the same pull up configuration than for the gpio */
+	if (gpio->dt_flags & GPIO_PULL_UP)
+		res = stm32_pwr_irq_set_pull_config(itr_num, WKUP_PULL_UP);
+	else if (gpio->dt_flags & GPIO_PULL_DOWN)
+		res = stm32_pwr_irq_set_pull_config(itr_num, WKUP_PULL_DOWN);
+	else
+		res = stm32_pwr_irq_set_pull_config(itr_num, WKUP_NO_PULL);
+	if (res) {
+		gpio_put(gpio);
+		return res;
+	}
+
+	if (IS_ENABLED(CFG_STM32_EXTI))
+		stm32_exti_set_tz(priv->exti, PWR_EXTI_WKUP1 + itr_num);
+
+	itr_desc_p->chip = &stm32mp2_pwr_itr_chip;
+	itr_desc_p->itr_num = itr_num;
+
+	return TEE_SUCCESS;
+}
+
 static TEE_Result stm32mp25_pwr_itr_add(const void *fdt, int wp_node,
 					struct itr_handler *hdl)
 {
@@ -377,11 +591,12 @@ stm32mp25_pwr_itr_alloc_add(const void *fdt, int wp_node, size_t it,
 	return res;
 }
 
-TEE_Result
-stm32mp25_pwr_irq_probe(const void *fdt, int node, int interrupt)
+TEE_Result stm32mp25_pwr_irq_probe(const void *fdt, int node)
 {
 	TEE_Result res = TEE_ERROR_GENERIC;
 	struct pwr_wkup_data *priv = NULL;
+	struct itr_chip *itr_chip = NULL;
+	size_t itr_num = DT_INFO_INVALID_INTERRUPT;
 
 	VERBOSE_PWR("Init PWR IRQ");
 
@@ -397,14 +612,17 @@ stm32mp25_pwr_irq_probe(const void *fdt, int node, int interrupt)
 			goto err;
 	}
 
-	res = interrupt_alloc_add_handler(interrupt_get_main_chip(),
-					  interrupt, pwr_it_handler,
-					  ITRF_TRIGGER_LEVEL, pwr_wkup_d,
-					  &priv->gic_hdl);
+	res = interrupt_dt_get(fdt, node, &itr_chip, &itr_num);
+	if (res)
+		panic("No interrupt defined in PWR");
+
+	res = interrupt_create_handler(itr_chip, itr_num, pwr_it_handler,
+				       pwr_wkup_d, ITRF_TRIGGER_LEVEL,
+				       &priv->parent_hdl);
 	if (res)
 		panic("Could not get wake-up pin IRQ");
 
-	interrupt_enable(interrupt_get_main_chip(), priv->gic_hdl->it);
+	interrupt_enable(priv->parent_hdl->chip, priv->parent_hdl->it);
 
 	if (IS_ENABLED(CFG_CORE_ASYNC_NOTIF))
 		notif_register_driver(&stm32_pwr_notif);
-- 
2.47.2

